addons/mrp/models/mrp_production.py

```diff
@@ -2164,40 +2170,50 @@ def _link_bom(self, bom):
         def operation_key_values(record):
             return tuple(record[key] for key in ('company_id', 'name', 'workcenter_id'))
 
-        product_attribute_ids = self.product_id.product_template_attribute_value_ids.ids
-        ratio = bom.product_qty / self.product_qty
-
-        bom_lines, bom_byproducts, operations = [], [], []
-        for bom_line in bom.bom_line_ids:
-            if not bom_line.bom_product_template_attribute_value_ids or\
-               any(att_val.id in product_attribute_ids for att_val in bom_line.bom_product_template_attribute_value_ids):
-                bom_lines.append(bom_line)
-
-        for bom_byproduct in bom.byproduct_ids:
-            if not bom_byproduct.bom_product_template_attribute_value_ids or\
-               any(att_val.id in product_attribute_ids for att_val in bom_byproduct.bom_product_template_attribute_value_ids):
-                bom_byproducts.append(bom_byproduct)
+        def filter_records(record_set):
+            product_attribute_ids = self.product_id.product_template_attribute_value_ids.ids
+            records_by_id = {}
+            for record in record_set:
+                if not record.bom_product_template_attribute_value_ids or\
+                any(att_val.id in product_attribute_ids for att_val in record.bom_product_template_attribute_value_ids):
+                    records_by_id[record.id] = record
+            return records_by_id
 
-        for operation in bom.operation_ids:
-            if not operation.bom_product_template_attribute_value_ids or\
-               any(att_val.id in product_attribute_ids for att_val in operation.bom_product_template_attribute_value_ids):
-                operations.append(operation)
+        ratio = self._get_ratio_between_mo_and_bom_quantities(bom)
+        bom_lines_by_id = filter_records(bom.bom_line_ids)
+        bom_byproducts_by_id = filter_records(bom.byproduct_ids)
+        operations_by_id = filter_records(bom.operation_ids)
```

Comment:
is there any reason why this wasn't in the original commit rather than modifying another commit in the same PR?

Reply:
Yes, I did one commit by feature. The first one add the generate BoM things, the second one add the update BoM code in `mrp`.
I try to separate as much as possible every feature since if we want only the generate BoM part, we can without modifying the code

Reply:
Ah, i mean it seems like you could have used `filter_by_attributes` in the first commit since it's doing the same logic with less code, just needs to be called not within a dict unless I'm reading it wrong :thinking:

Reply:
Ok !
You're right, I made the change, thanks :)